### ISSUE TITLE
Future of vim.str_{byte,utf}index

### ISSUE THREAD

**Issue by tris203** [neovim/neovim#30804]
### Problem

# vim.str_byteindex/ vim.str_utfindex
There are several limitations to the current implementation of these functions

1. They throw given out-of-range indexes

   >*lua-error-handling*
   >Lua functions may throw |lua-errors| for exceptional (unexpected) failures,
   >which you can handle with |pcall()|.
   >*lua-result-or-message*
   >When failure is normal and expected, it's idiomatic to return `nil` which
   >signals to the caller that failure is not "exceptional" and must be handled.

   An invalid (often user/lsp-provided) index doesn't feel exceptional to me and should be handled by the caller.

   `:h dev-error-patterns` suggests that `retval, errmsg` is the most preferable way to handle errors and I think that this is a good example of where this could be used
 
2. They cannot accept an encoding
    This leads to wrapper functions in the current LSP usageThis was identified as a potential improvement in #25272

3. The current signatures are not extensible</summary>
    * Calling: The function currently uses a fairly opaque boolean value to indicate to identify the encoding
    * Returns: The fact it can throw means wrapping in pcalls.
 
4. The current naming doesn't follow suggestions in `:h dev-naming` and I think `get` would be suitable

### Common

To change the signatures to this level this would be a breaking change. And as such I would propose the following:

Create new functions of ```vim.str_getbyteindex()``` and ```vim.str_getutfindex()```

These functions would both take a ```string, opts``` signature where opts is a table with the following keys:
* ```encoding?``` - The encoding to use, defaulting to utf-16
* `index?` - The string index to find, defaulting to the length of the string

These functions should return ```retval, errmsg```, returning the index or nil. This would make the LSP handling code simpler

### vim.str_getbyteindex()

Proposed signature:
```lua
---@param str string
---@param opts { index?: integer, encoding?: "utf-8" | "utf-16" | "utf-32" }
---@return integer? # UTF-8 translated byte index
---@return string? # error messsage, if applicable
function vim_str_getbyteindex(str, opts) end
```
This means the LSP code could be as follows
```lua
return vim.str_getbyteindex(str, { index = 1, encoding = 'utf-16'}) or #str
```

### vim.str_getutfindex()
Proposed signature:
```lua
---@param str string
---@param opts { index?: integer, encoding?: "utf-8" | "utf-16" | "utf-32" }
---@return integer? # Translated UTF-8 index
---@return string? # error message, if applicable
function vim_str_getutfindex(str, opts) end
```

`str_utfindex()` currently has multiple return values for each encoding. Although this appears presently to just be a bit of a dance to get the right one from the return. I can't see anywhere at least in the Neovim codebase where this would result in needing two function calls

## Questions:
* Should the index be in the opts table? or should it be kept as the second argument?
* What would the timeline be for deprecating the existing functions (as far as the Neovim code based they only seem to be used in `vim.region` (already deprecated) and the LSP code. But they are used in the plugin ecosystem.

**Comment by lewis6991** [comment]
> An invalid (often user/lsp-provided) index doesn't feel exceptional to me and should be handled by the caller.

It is exceptional because it is the result of invalid inputs and not caused by any external factors. This is unlike functions in vim.uv which call out to the system and can error for various reasons which can't be predicted by the caller. In this case it is trivial for the caller to check beforehand if it is out of index.

Do we know why the LSP util function checks this beforehand instead of throwing? Were there cases of responses from LSP returning incorrect indexes or was it added to mask other issues in the subsystem? In either case it is worth logging when this happens so either can be fixed properly.

I also don't think we need to add an `opts` argument. This isn't the sort of function that needs extensibility. This is the kind of function that benefits most by being simple.

**Comment by tris203** [comment]
> > An invalid (often user/lsp-provided) index doesn't feel exceptional to me and should be handled by the caller.
> 
> It is exceptional because it is the result of invalid inputs and not caused by any external factors. This is unlike functions in vim.uv which call out to the system and can error for various reasons which can't be predicted by the caller. In this case it is trivial for the caller to check beforehand if it is out of index.
> 
It is, but it requires either a wrapper function or this to be done in each call. 
> Do we know why the LSP util function checks this beforehand instead of throwing? Were there cases of responses from LSP returning incorrect indexes or was it added to mask other issues in the subsystem? In either case it is worth logging when this happens so either can be fixed properly.
> 
There were some issues, but is also what the spec asks for.
```
    -- LSP spec: if character > line length, default to the line length.
    -- https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#position
```
> I also don't think we need to add an `opts` argument. This isn't the sort of function that needs extensibility. This is the kind of function that benefits most by being simple.

The function isn't getting anymore complex, its just standardising the way the arguments are passed in line with the rest of the APi.
The code was implemented here #10688. Since 2019 there has been a lot of standardisation of the API with naming/return conventions. This is simply a proposal to bring this function inline with those standards.

**Comment by lewis6991** [comment]
> It is, but it requires either a wrapper function or this to be done in each call.

Then add a wrapper, just ensure it remains private. That's going to be much more preferable to any breaking change, especially when such change isn't obviously an improvement.

> The function isn't getting anymore complex

Adding an `opts` field does make it more complex as every `opts` field requires a table allocation.

It is not the intention for every Nvim API function to contain an `opts` field, especially simple functions that are often called in a tight loop.

**Comment by lewis6991** [comment]
I just looked at this a bit more closely, and the best path forward is less obvious.

- Both functions return a (UTF-16, UTF-32) tuple. So no matter what we do, changing these functions will be a breaking change, unless we incorporate some overloading.

- `_str_utfindex_enc` requires 2 calls to `str_utfindex` for the UTF-16/32 cases. This would be nice to avoid.

So despite my past comments, I actually think what you have in your PR is pretty close to what we should go for. But personally I think we should drop the `opts` route and just make everything a positional argument.

```lua
local col = vim.str_utfindex(line, encoding, [index, [, no_out_of_range]])
```

We can support the old versions by dispatching on the type of argument 2, and deprecating that form.

```lua
vim.str_utfindex(line)                            -- (utf-32 length, utf-16 length) - deprecated (can make it rerturn utf-16 length in he future)
vim.str_utfindex(line, index)                     -- (utf-32 index, utf-16 index) - deprecated
vim.str_utfindex(line, 'utf-16')                  -- utf-16 length
vim.str_utfindex(line, 'utf-16', index)           -- utf-16 index
vim.str_utfindex(line, 'utf-16', math.huge)       -- error: index out of range 
vim.str_utfindex(line, 'utf-16', math.huge, true) -- utf-16 length
```

This will mean introducing some return type polymorphism (which I hate), but since the other form will be deprecated, we can prefer the new return types for the type annotations.

@justinmk thoughts?

**Comment by justinmk** [comment]
> This will mean introducing some return type polymorphism (which I hate), but since the other form will be deprecated, we can prefer the new return types for the type annotations.

Yes that sounds good. 

So work will continue in https://github.com/neovim/neovim/pull/30735 and the other PR can be closed?

**Comment by tris203** [comment]
> > This will mean introducing some return type polymorphism (which I hate), but since the other form will be deprecated, we can prefer the new return types for the type annotations.
> 
> Yes that sounds good.
> 
> So work will continue in [#30735](https://github.com/neovim/neovim/pull/30735) and the other PR can be closed? Hard to follow the delta between them.

I think that the only difference is if we wrap the C functions or go for a pure Lua implementation. 

I think wrapping the existing code is simpler/cleaner for PR reviews, so we can revisit the Lua implementation later if necessary. I will close mark other PRs as draft for now

**Comment by lewis6991** [comment]
> I think that the only difference is if we wrap the C functions or go for a pure Lua implementation.

Yes, it's good practice to separate implementation changes separate from API changes when possible.

**Comment by justinmk** [comment]
Most of this was resolved by https://github.com/neovim/neovim/pull/30735. If there is remaining open question, track it in a new issue.
